File: Yr2022/day_two.py
from time import perf_counter_ns
from enum import IntEnum

class RPS(IntEnum):
    ROCK = 1
    PAPER = 2
    SCISSORS = 3

class OUTCOME(IntEnum):
    WIN = 6
    DRAW = 3
    LOSE = 0

move_map = {'A':RPS.ROCK, 'B':RPS.PAPER, 'C': RPS.SCISSORS,
            'X':RPS.ROCK, 'Y':RPS.PAPER, 'Z':RPS.SCISSORS}

win_map = {RPS.ROCK:RPS.SCISSORS, RPS.PAPER:RPS.ROCK, RPS.SCISSORS:RPS.PAPER}

lose_map = {RPS.SCISSORS:RPS.ROCK, RPS.ROCK:RPS.PAPER, RPS.PAPER:RPS.SCISSORS}

outcome_map = {'X':OUTCOME.LOSE, 'Y':OUTCOME.DRAW, 'Z':OUTCOME.WIN}


def strategy_score(opponent_move, my_move):
    """Calculate score assuming 2nd column is my move"""
    result_score = my_move
    if opponent_move == my_move:               #Draw
        result_score += OUTCOME.DRAW
    elif win_map[my_move] == opponent_move:    #Win
        result_score += OUTCOME.WIN
    return result_score


def outcome_score(opponent_move, outcome_required):
    """Calculate score assuming 2nd column is outcome required"""
    result_score = outcome_required
    if outcome_required == OUTCOME.DRAW:
        result_score += opponent_move
    elif outcome_required == OUTCOME.LOSE:
        result_score += win_map[opponent_move]
    else:
        result_score += lose_map[opponent_move]
    return result_score


def calculate_score(filename, outcome_method=False):
    """Calculate score using one of two methods"""
    score_counter = 0
    with open(filename, 'r') as file:
        for line in file:
            # I'm not thrilled with using magic numbers for indices, not very robust if input is wrong
            #moves = [move_map[move] for move in line.replace('\n','').split(' ')]
            if outcome_method:                
                opponent_move, outcome_required = move_map[line[0]], outcome_map[line[2]]
                score_counter += outcome_score(opponent_move, outcome_required)            
            else:
                opponent_move, my_move = move_map[line[0]], move_map[line[2]]
                score_counter += strategy_score(opponent_move, my_move)
    return score_counter


if __name__ == "__main__":
    time_start = perf_counter_ns()
    file_loc = r"C:\Users\Chris\Documents\CodeProjects\AdventOfCode\Yr2022\PuzzleData\puzzle_day_two.txt"
    print(calculate_score(file_loc, True))
    print(f"Elapsed time : {(perf_counter_ns() - time_start)/1000:.0f} microseconds")

File: Yr2022/src/day_two.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;
use std::time::Instant;
use std::collections::HashMap;

#[macro_use]
extern crate lazy_static;

#[derive(Eq, PartialEq, Hash, Copy, Clone)]
pub enum RPS {
    ROCK = 1,
    PAPER = 2,
    SCISSORS = 3,
}

#[derive(Eq, PartialEq, Hash, Copy, Clone)]
pub enum OUTCOME {
    WIN = 6,
    DRAW = 3,
    LOSE = 0,
}

lazy_static! {
static ref MOVE_MAP : HashMap<char, RPS> = HashMap::from([('A',RPS::ROCK), ('B',RPS::PAPER), ('C',RPS::SCISSORS),
            ('X',RPS::ROCK), ('Y',RPS::PAPER), ('Z',RPS::SCISSORS)]);

static ref WIN_MAP : HashMap<RPS, RPS> = HashMap::from([(RPS::ROCK,RPS::SCISSORS), (RPS::PAPER,RPS::ROCK), (RPS::SCISSORS,RPS::PAPER)]);

static ref LOSE_MAP : HashMap<RPS, RPS> = HashMap::from([(RPS::SCISSORS,RPS::ROCK), (RPS::ROCK,RPS::PAPER), (RPS::PAPER,RPS::SCISSORS)]);

static ref OUTCOME_MAP : HashMap<char, OUTCOME> = HashMap::from([('X',OUTCOME::LOSE), ('Y',OUTCOME::DRAW), ('Z',OUTCOME::WIN)]);
}


fn main() {   
    let start = Instant::now();
    let file_name= "C:\\Users\\Chris\\Documents\\CodeProjects\\AdventOfCode\\Yr2022\\PuzzleData\\puzzle_day_two.txt";
    let tot_score = sum_scores(file_name, true);
    let duration = start.elapsed();
    println!("Total score: {}", tot_score);    
    println!("Time taken to execute: {} microseconds", duration.as_micros());    
}


fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where P: AsRef<Path>, {
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}


fn strategy_score(opponent_move: &RPS, my_move : &RPS) -> i32 {
    //Calculate score assuming 2nd column is my move
    let mut result_score = *my_move as i32;
    let win_move = WIN_MAP.get(my_move).unwrap();
    if opponent_move == win_move {
        result_score += OUTCOME::WIN as i32; }
    else if opponent_move == my_move {
        result_score += OUTCOME::DRAW as i32;        
    }
    return result_score;
}


fn outcome_score(opponent_move: &RPS, outcome_required : &OUTCOME) -> i32{
    //Calculate score assuming 2nd column is outcome required
    let mut result_score = *outcome_required as i32;
    match outcome_required {
        OUTCOME::DRAW => result_score += *opponent_move as i32,
        OUTCOME::LOSE => result_score += *WIN_MAP.get(opponent_move).unwrap() as i32,
        OUTCOME::WIN => result_score += *LOSE_MAP.get(opponent_move).unwrap() as i32,
    }
    return result_score;
}


fn sum_scores(my_file_name: &str, outcome_method: bool) -> i32 {
    // Sum the scores using one of two methods
    let mut total_score: i32 = 0;
    if let Ok(lines) = read_lines(my_file_name) {
        for line in lines {
            if let Ok(ip) = line {
                let first_char = ip.chars().nth(0).unwrap();
                let second_char = ip.chars().nth(2).unwrap();
                if outcome_method {
                    total_score += outcome_score(MOVE_MAP.get(&first_char).unwrap(),
                                                 OUTCOME_MAP.get(&second_char).unwrap());
                }
                else {
                    total_score += strategy_score(MOVE_MAP.get(&first_char).unwrap(),
                                                 MOVE_MAP.get(&second_char).unwrap());
                }            
            }
        }
    }
    return total_score
}

